Added ability to GET, PUT and POST to the Xero Payroll API

gazzenger opened the pull request:
Thanks so much for this fantastic module.

I made a minor change so that a user can now also perform GET, POST and PUT commands to the Payroll API on Xero. This means users can now handle timesheets, payitems as well as many other features.

I've left support for the command xero.call, however I've added the following commands,
xero.callAccounting() and
xero.callPayroll()

These commands work the same way as xero.call, but allow distinguishing between Accounting API (which is what the current code does), and the Payroll API.

An example to get a list of timesheets would be,
xero.callPayroll('GET', "/Timesheets", null, function (err, json) {...})

Let me know your thoughts on the additions,

Kind Regards,
Gary

gnamestnik committed:
Added the ability to query both the Accounting API and the Payroll API of Xero by adding two functions xero.callPayroll() and xero.callAccounting().

gnamestnik committed:
Updated readme.md file

gnamestnik committed:
Fixed typo in Readme.md

presolve commented:
Thanks for doing this!  I am half way through writing something to achieve the same thing and thought i should look! lol  I had a look at your update and am wondering if there is there a reason you don't just just switch the XERO_API_URL value based on the 'path' received?  That would save having to specify which endpoint you want to call and would be seemless (?).  What you think?

presolve commented:
Something like...

`var paths = {
    "accounting": ["/Contacts"],
    "payroll": ["/Employees"]
}

function contains(a, obj) {
    for (var i = 0; i < a.length; i++) {
        if (a[i] === obj) {
            return true;
        }
    }
    return false;
}

function Xero(key, secret, rsa_key, showXmlAttributes, customHeaders, version) {
    this.key = key;
    this.secret = secret;

    this.parser = new xml2js.Parser({explicitArray: false, ignoreAttrs: showXmlAttributes !== undefined ? (showXmlAttributes ? false : true) : true, async: true});

    this.oa = new oauth.OAuth(null, null, key, secret, '1.0', null, "PLAINTEXT", null, customHeaders);
    this.oa._signatureMethod = "RSA-SHA1"
    this.oa._createSignature = function(signatureBase, tokenSecret) {
        return crypto.createSign("RSA-SHA1").update(signatureBase).sign(rsa_key, output_format = "base64");
    }
}

Xero.prototype.call = function(method, path, body, callback) {
    var self = this;
    
    XERO_API_URL = XERO_BASE_URL + ((contains(paths.payroll,path)) ? ('/payroll.xro/1.0') : ('/api.xro/2.0'));

    var post_body = null;`

gazzenger commented:
Heya,
The reason I haven't switched the XERO_API_URL on the value based on a path, was it simple when specifying either the Accounting or Payroll APIs. But I wanted to keep it backwards compatible with what you've written.
I'm really happy for anything either way :)

gazzenger commented:
Yep I think that looks good.

On Wed, Nov 23, 2016 at 2:55 PM, presolve <notifications@github.com> wrote:

> Something like...
>
> `var paths = {
> "accounting": ["/Contacts"],
> "payroll": ["/Employees"]
> }
>
> function contains(a, obj) {
> for (var i = 0; i < a.length; i++) {
> if (a[i] === obj) {
> return true;
> }
> }
> return false;
> }
>
> function Xero(key, secret, rsa_key, showXmlAttributes, customHeaders,
> version) {
> this.key = key;
> this.secret = secret;
>
> this.parser = new xml2js.Parser({explicitArray: false, ignoreAttrs: showXmlAttributes !== undefined ? (showXmlAttributes ? false : true) : true, async: true});
>
> this.oa = new oauth.OAuth(null, null, key, secret, '1.0', null, "PLAINTEXT", null, customHeaders);
> this.oa._signatureMethod = "RSA-SHA1"
> this.oa._createSignature = function(signatureBase, tokenSecret) {
>     return crypto.createSign("RSA-SHA1").update(signatureBase).sign(rsa_key, output_format = "base64");
> }
>
> }
>
> Xero.prototype.call = function(method, path, body, callback) {
> var self = this;
>
> XERO_API_URL = XERO_BASE_URL + ((contains(paths.payroll,path)) ? ('/payroll.xro/1.0') : ('/api.xro/2.0'));
>
> var post_body = null;`
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/thallium205/xero/pull/23#issuecomment-262444989>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKDdNXIJ9TV6iVyJt2SCGhWmiplxDGgnks5rA-NMgaJpZM4Kg5E6>
> .
>



-- 

*Gary T. Namestnik*



Mobile: +61 (0) 4 1537 0736

presolve commented:
I'm a git n00b... sorry... I've made this patch which is working great for me now.  It's dynamic and works for both US and AUS payroll paths.   Any chance you could put this forward?

[index.txt](https://github.com/thallium205/xero/files/608262/index.txt)

gazzenger commented:
Same here,
I'll have a look at adding a new commit to my branch so hopefully you'll see it in the pull request.
I might do it later this afternoon if that's alright.

gazzenger commented:
Hey I've been looking over the code, and I noticed that you've got an array explicitly listing the URIs for using the payroll API,
```
var paths = {
    'payroll': [
        '/Employees',
        '/Leaveapplications',
        '/Payitems',
        ...
```
And then to determine if the user intends to use the payroll or accounting API, you check if the path is contained in the array paths.payroll.

`XERO_API_URL = XERO_BASE_URL + ((contains(paths.payroll,path)) ? ('/payroll.xro/1.0') : ('/api.xro/2.0'));`

Although this seems like a good method, I can spot that at least one URI has the same name in both the accounting and payroll API.

So you can have endpoints such as these:
```
https://api.xero.com/payroll.xro/1.0/employees 
https://api.xero.com/api.xro/2.0/employees
```
I'm thinking the best way to avoid this issue, could be to add an extra parameter when calling the xero module, and specify which API to use, maybe set the default to be accounting.

`xero(key, secret, rsa_key, showXMLAttrib, customHeaders, **apiType**)`

Let me know your thoughts

presolve commented:
Yep... I literally only 5 mins ago came across that! lol  It's a bummer cos the Employees path is the ONLY one that exists in both endpoints and 99% of the time it's not even used (most people use the Employees Payroll endpoint.  I originally started down the "mode" switch path so I'll just redo it with the switch instead. :)
Thanks for reviewing!!

presolve commented:
i'll learn git so i can push a patch i swear... but here's the update... now it'll work for all endpoints and you just specify the 'endpoint'.

```
    switch (endpoint) {
        case 'accounting':
            endpoint = '/api.xro/2.0';
            break;
        case 'payroll':
            endpoint = '/payroll.xro/1.0';
            break;
        case 'assets':
            endpoint = '/assets.xro/1.0';
            break;
        case 'files':
            endpoint = '/files.xro/1.0';
            break;
        default:
            endpoint = '/api.xro/2.0';
    }
    XERO_API_URL = XERO_BASE_URL + endpoint;
```
So when you call it now you just

`xero.call('GET', 'payroll', '/Employees', null, function(err, xeroEmployees) {...`

What do you think? (complete file attached... just rename to index.js... :)
[index.txt](https://github.com/thallium205/xero/files/610367/index.txt)

gazzenger commented:
I love it, what're the assets and files endpoints used for?

presolve commented:
Uploading files and such! :P
https://developer.xero.com/documentation/files-api/files/

I think i'll start using this feature now that it's working! :D

gazzenger commented:
Do you want me to do the updates in my branch?

presolve commented:
if you wouldn't mind... that would be great. :)

thallium205 commented:
Whenever we feel ready I can get this change published to npm.

On Wed, Nov 23, 2016 at 7:20 PM, presolve <notifications@github.com> wrote:

> if you wouldn't mind... that would be great. :)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/thallium205/xero/pull/23#issuecomment-262681507>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1RgPkGgWZ3mGJxTCtom1tqaZTDSTaYks5rBQJ1gaJpZM4Kg5E6>
> .
>

presolve commented:
Sooner the better!  I'm using it now and it's working great. :)
Thanks heaps!!!

gazzenger commented:
Sorry, I was out at lunch, let me update it in my branch (I think it'll automatically go through with the pull request)

gnamestnik committed:
Updated the index.js file by removing the two external functions xero.callAccounting and xero.callPayroll, and replaced them with an endpoint parameter in the main function xero.call(method, endpoint, path, body, callback)
The endpoint takes the various strings, payroll, accounting, assets, files, and the default is the accounting.
Also added these to the md file.

gazzenger commented:
Alrighty, it should all be updated now.
See what you think

presolve reviewed (approved):
Looks great!

gazzenger commented:
I've given the new module a whirl, and it works well

presolve commented:
Same!  Thanks heaps for your help @gazzenger
And thanks @thallium205 for the module!!

gazzenger commented:
@thallium205 The pull request is ready now

gnamestnik committed:
Fixed bug with the payitems section

Commit:
updated package.json

Commit:
New additions to package.json

Commit:
Update to README file

Commit:
Updated version number

thallium205 commented:
The branch has got some conflicts and can't be merged.  If you resolve I'll merge. Thanks!
